### PR TITLE
[Tune] Add support for nested hyperparams in PB2

### LINKED ISSUE #29786: [Tune] PB2 is missing unit tests
### Description

PB2 is only tested via examples in `tune/examples` currently. We should add unit tests checking that PB2 scheduling behavior is correct.

See https://github.com/ray-project/ray/pull/29102 for an example of what should be tested.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

This PR enables nested passing hyperparameters for the PB2 scheduler.

This PR also makes a few minor improvements to PB2 (happy to separate out these changes if needed):
1. Hyperparameter initialization (if missing from param space) should be sampled uniformly between bounds. Currently, PB2 falls back to PBT for sampling initial hyperparameters, which will just choose between the low/high values.
2. Allow `custom_explore_fn` to be passed into PB2 to match PBT functionality. This solves a user request here: https://discuss.ray.io/t/pb2-hyper-parameters-as-integers/8822.

## Related issue number

Closes https://github.com/ray-project/ray/issues/30124
Closes https://github.com/ray-project/ray/issues/29786

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/tests/test_trial_scheduler_pbt.py [hunk 1]
```diff
@@ -778,7 +783,162 @@ def test_config(
...
+    assert trial.config["b"]["c"] == 3.0
+
+
+def test_pb2_hyperparam_bounds_validation():
+    hyperparam_bounds = {"a": [1.0, 2.0], "b": {"c": [2.0, 4.0, 6.0]}}
+    with pytest.raises(ValueError):
+        create_pb2_scheduler(hyperparam_bounds=hyperparam_bounds)
+
+
+def test_pb2_custom_explore_fn():
```
Comment (Yard1) [inline python/ray/tune/tests/test_trial_scheduler_pbt.py:924]: can we also check if the scheduler can be serialised if the function is set to eg. a lambda?

**Comment by Yard1** [comment]
@justinvyu

File: python/ray/tune/tests/test_trial_scheduler_pbt.py [hunk 2]
```diff
@@ -778,7 +783,162 @@ def test_config(
         )
 
 
-if __name__ == "__main__":
-    import pytest
+def create_pb2_scheduler(
```
Comment (krfricke) [inline python/ray/tune/tests/test_trial_scheduler_pbt.py:786]: Can we make these functions internal?

And maybe add typehints

**Review by krfricke — COMMENTED** [review]
Logic looks good to me, should we aim to ship this PR soon?

**Comment by justinvyu** [comment]
Will aim to get it wrapped up this week!

**Review by krfricke — APPROVED** [review]
Thanks!
